### PR TITLE
login pr

### PR DESCRIPTION
**Pull request by Skr1pt1k**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,4 +1,16 @@
 # frozen_string_literal: true
 
 class ApplicationController < ActionController::API
+  def authorize_request
+    header = request.headers['Authorization']
```
Comment (alex-pro) [inline app/controllers/application_controller.rb]: use dig and & to refactor

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,4 +1,16 @@
...
+    header = header.split(' ').last if header
+    begin
+      @decoded = JsonWebToken.decode(header)
+      @current_user = User.find(@decoded[:user_id])
+    rescue ActiveRecord::RecordNotFound
+      render json: { error: 'unauthorized' }, status: :unauthorized
+    rescue JWT::DecodeError
+      render json: { error: 'unauthorized' }, status: :unauthorized
```
Comment (alex-pro) [inline app/controllers/application_controller.rb]: duplication

File: app/controllers/authentication_controller.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      time = Time.now + 1.hours.to_i
+      render json: { token: token, exp: time.strftime('%m-%d-%Y %H:%M') }, status: :ok
+    else
+      render json: { error: 'unauthorized' }, status: :unauthorized
+    end
+  end
+
+  private
+
+  def login_params
```
Comment (alex-pro) [inline app/controllers/authentication_controller.rb]: where is it used?

File: app/controllers/authentication_controller.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class AuthenticationController < ApplicationController
+  before_action :authorize_request, except: :create
+
+  def create
+    user = User.find_by_email(params[:email])
```
Comment (alex-pro) [inline app/controllers/authentication_controller.rb]: move it to expose

File: lib/json_web_token.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+class JsonWebToken
+  SECRET_KEY = Rails.application.secrets.secret_key_base. to_s
```
Comment (alex-pro) [inline lib/json_web_token.rb]: strange space symbol in the end

Comment (alex-pro) [inline lib/json_web_token.rb]: and is it necessary to convert it to a string? what kind of class this secret key base is?

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -2,10 +2,13 @@
...
 
-  def authorize_request
-    @jwt = cookies.signed[:jwt]
-    @decoded = JsonWebToken.decode(@jwt)
-    @current_user = User.find(@decoded[:user_id])
+  def current_user
+    if cookies.signed[:jwt].present?
+      jwt = cookies.signed[:jwt]
+      decoded = JsonWebToken.decode(jwt)
+      @current_user = User.find(decoded[:user_id])
```
Comment (Skr1pt1k) [inline app/controllers/application_controller.rb]: i forget to delete this - '@' , in next push...

Comment (alex-pro) [reply app/controllers/application_controller.rb]: cookies.signed[:jwt] to var

Comment (Skr1pt1k) [reply app/controllers/application_controller.rb]: ok

File: app/models/todo_list.rb [hunk 7]
```diff
@@ -2,6 +2,11 @@
 
 class TodoList < ApplicationRecord
   has_many :records, dependent: :destroy
+  belongs_to :user
 
   validates :title, presence: true, length: { minimum: 3 }
+
+  def self.by_user(user)
+    TodoList.where(user: user)
```
Comment (alex-pro) [inline app/models/todo_list.rb]: move to scope
also remove TodoList

File: app/models/user.rb [hunk 8]
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 class User < ApplicationRecord
+  has_many :todo_lists
```
Comment (alex-pro) [inline app/models/user.rb]: don't forget about dependent destroy

File: config/application.rb [hunk 9]
```diff
@@ -24,10 +24,15 @@ class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 5.2
 
+    config.middleware.use ActionDispatch::Cookies
+
     config.middleware.insert_before 0, Rack::Cors do
       allow do
-        origins '*'
-        resource '*', headers: :any, methods: :any
+        origins 'http://localhost:8080'
```
Comment (alex-pro) [inline config/application.rb]: move it to .env

File: app/controllers/records_controller.rb [hunk 10]
```diff
@@ -1,33 +1,48 @@
...
   def destroy
+    authorize record
+
     record.destroy
     head :no_content, status: :ok
   end
 
   private
 
+  def bad_request_params
```
Comment (Skr1pt1k) [inline app/controllers/records_controller.rb]: Already renamed to bad_request_render, will see in next pr

**Comment by Skr1pt1k** [comment]
Working with rails services

File: app/controllers/records_controller.rb [hunk 11]
```diff
@@ -1,33 +1,48 @@
...
+    render json: record.errors, status: :unprocessable_entity
+  end
+
   def record_params
     params.require(:record).permit(:title, :done, :date, :details, :priority)
```
Comment (Skr1pt1k) [inline app/controllers/records_controller.rb]: already delete ':date', will see in next pr

File: config/environments/development.rb [hunk 12]
```diff
@@ -44,8 +44,8 @@
     address: 'smtp.gmail.com',
```
Comment (Skr1pt1k) [inline config/environments/development.rb]: should i have hide address, -> .env?

File: spec/controllers/authentications_controller_spec.rb [hunk 13]
```diff
@@ -3,4 +3,25 @@
...
+      it { expect(json['user']).to eq(user.email) }
+    end
+
+    context 'when failed' do
+      subject(:create) { post :create, params: invalid_params }
+
+      let(:invalid_params) { { email: user.email, password: 'asdasdasd' } }
+
+      it { expect(create).to have_http_status(:unauthorized) }
+      it { expect(json['user']).not_to eq(user.email) }
```
Comment (alex-pro) [inline spec/controllers/authentications_controller_spec.rb]: better to check your real response body. { error: 'unauthorized' } does not contain email key

**Pull request merged**